src/conn/conn_dhandle.c

```diff
@@ -708,6 +708,15 @@ __wt_conn_dhandle_discard(WT_SESSION_IMPL *session)
 	__wt_session_close_cache(session);
 	F_SET(session, WT_SESSION_NO_DATA_HANDLES);
 
+	/*
+	 * The connection may have an open metadata cursor handle. We cannot
+	 * close it before now because it's potentially used when discarding
+	 * other open data handles. Close it before discarding the underlying
+	 * metadata handle.
+	 */
+	if (session->meta_cursor != NULL)
+		WT_TRET(session->meta_cursor->close(session->meta_cursor));
+
```

Comment:
> Where do we close the new cursor when closing a session handle?

The metadata cursor is on the cursor list, so we close it in `__session_close()`:
```
/* Close all open cursors. */
while ((cursor = TAILQ_FIRST(&session->cursors)) != NULL) {
        /*
         * Notify the user that we are closing the cursor handle
         * via the registered close callback.
         */
        if (session->event_handler->handle_close != NULL &&
            !WT_STREQ(cursor->internal_uri, WT_LAS_URI))
                WT_TRET(session->event_handler->handle_close(
                    session->event_handler, wt_session, cursor));
        WT_TRET(cursor->close(cursor));
}
```
Or is that not what you're asking?